db/seeds.rb

```diff
@@ -60,34 +61,88 @@
   setting.value = ModelConfig.generation.default.last&.id
 end
 
-# DEV
+# PROVISIONING
 #
-if Rails.env == "development"
-  ModelServer.find_or_create_by!(name: "localhost") do |ms|
-    ms.url = model_endpoint
-    ms.provider = "ollama"
-  end
+dev_model_servers = [
+  {
+    name: "localhost",
+    provider: "ollama",
+  }
+]
+
+dev_model_configs = [
+  {
+    name: "mistral:instruct",
+    model_server: "localhost",
+    model: "mistral:instruct",
+    temperature: 0.1,
+  },
+  {
+    name: "gemma:7b",
+    model_server: "localhost",
+    model: "gemma:7b",
+    temperature: 0.2,
+  },
+  {
+    name: "nomic-embed-text",
+    model_server: "localhost",
+    model: "nomic-embed-text",
+    embedding: true,
+  }
+]
+
+provisioned_model_servers = if ENV["PROVISIONED_MODEL_SERVERS"].present?
+  JSON.parse(ENV["PROVISIONED_MODEL_SERVERS"])
+elsif Rails.env == "development"
+  dev_model_servers
+else
+  []
+end
 
-  ModelConfig.find_or_create_by!(name: "mistral:instruct", model_server: ModelServer.first) do |mc|
-    mc.model = "mistral:instruct"
-    mc.temperature = 0.1
-  end
+provisioned_model_configs = if ENV["PROVISIONED_MODEL_CONFIGS"].present?
+  JSON.parse(ENV["PROVISIONED_MODEL_CONFIGS"])
+elsif Rails.env == "development"
+  dev_model_configs
+else
+  []
+end
 
-  ModelConfig.find_or_create_by!(name: "gemma:7b", model_server: ModelServer.first) do |mc|
-    mc.model = "gemma:7b"
-    mc.temperature = 0.2
-  end
 
-  ModelConfig.find_or_create_by!(name: "all-minilm", model_server: ModelServer.first) do |mc|
-    mc.model = "all-minilm"
-    mc.embedding = true
-  end
+provisioned_model_servers_names = provisioned_model_servers.map { |ms| ms["name"] }
+provisioned_model_config_names = provisioned_model_configs.map { |mc| mc["name"] }
 
-  ModelConfig.find_or_create_by!(name: "nomic-embed-text", model_server: ModelServer.first) do |mc|
-    mc.model = "nomic-embed-text"
-    mc.embedding = true
-  end
+ModelConfig.where(provisioned: true).where.not(name: provisioned_model_config_names).update(available: false, provisioned: false)
+ModelServer.where(provisioned: true).where.not(name: provisioned_model_servers_names).update(available: false, provisioned: false)
+
+provisioned_model_servers.each do |config|
+  puts "provisioning model server for `#{config["name"]}` ..."
 
+  ModelServer.find_or_create_by!(name: config["name"]).update(
```

Comment:
This is nice, but has the same problem as the original that if any fields other than `name` are required (and they really should be, I'll fix that) this won't work.

In cases like this, where you want it to find OR basically call `.new` (which won't try to save to the DB and so won't invoke validations), use `find_or_initialize_by` chained with an `update`.

Reply:
I simplified this. It holds no hands and just assumes a given `hash` has the appropriate fields. That, plus the use of `update!` will offer some informative stack traces should the end user miss anything